accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+That means that the following commands would be equivalent:
+
+- `dotnet build -a x64 --os win`
+- `dotnet build -r win-x64 --no-self-contained`
+
+This asymmetry largely locks us into making the breaking change to `-r`. If we don't, the architecture targeting system will become even more confusing than it already is.
+
+Specifying both a RID and the short-hand syntax would be an error, for example with the following commands:
+
+- `dotnet build -r win-x64 -a arm64`
+- `dotnet build -r linux-arm64 --os win -a x64`
+
+The shorthand syntax is intended as a CLI and not MSBuild concept. For scenarios that today require specifying RIDs in project file, the intent is that users will continue to specify RIDs. If we find that there is a need for a shorthand RID syntax in project files, then we can consider extending this syntax to MSBuild.
+
+## Implementation timeline
+
+This proposal includes multiple breaking changes. They are impactful enough that we need to ensure a few key characteristics:
+
+- Breaks are introduced in an early preview of a new .NET version.
+- Breaks can be avoided by using existing gestures in a previous .NET version.
+- The breaks are configurable in some way.
+
+That leaves us with the following plan:
+
+- Add high-value additive changes for x64 emulation in .NET 6.
+- Add gestures that enable avoiding the .NET 7 breaking changes.
+- Make rest of changes, including all breaking changes, in .NET 7 Preview 1 or 2.
+- Publish plan before .NET 6 releases.
+
+The second point -- adding gestures that avoid future breaks -- is a key part of the plan. We also need to ensure that we update documentation and blogs to ensure that the users are aware of the changes that they should make when using .NET 6 in preparation for .NET 7 and later versions.
+
+### .NET 6
+
+The following changes should be included in .NET 6, motivated by the x64 emulation scenario. They are all additive and non-breaking.
+
+- Enable shorthand RID syntax. Add to the following verbs (priority order):
+   - `build`
+   - `publish`
+   - `tool install`
+   - `test`
+   - `run`
+   - `watch run`
+   - `watch test`
+- Do NOT enable the `-r` RID syntax on these same verbs. Doing so would cause a breaking change in .NET 7, which seems like a unforgivable mistake.
+- Enable implicit RID where there is one known choice:
+   - The macOS Arm64 SDK should treat pre .NET 6 versions as X64 by default.
+   - The Windows Arm64 SDK should treat pre .NET 5 versions as X64 by default.
+- Parity syntax between `dotnet build` and `dotnet publish` for pivoting between self-contained and framework-dependent modalities, for example:
+   - `dotnet build -r win-x64 --self-contained`
+   - `dotnet build -r win-x64 --self-contained true`
+   - `dotnet build -r win-x64 --self-contained false`
+   - `dotnet build -r win-x64 --no-self-contained`
+- Add warning when using `-r` without a `--self-contained` or `--no-self-contained`. The warning is for .NET 6+ apps only.
```

Comment:
For platforms that only support self-contained apps (or vice versa if that exists), it seems redundant to have to specify it.

Reply:
It does. Someone has to win ;)

More seriously, `-r` is an equally legit option for framework dependent apps and self-contained ones. IMO, it is incoherent to change the deployment mode as a result of specifying a RID.